### ISSUE TITLE
Add support for drawing anti-aliased lines and thick lines.

### ISSUE THREAD

**Issue by theotherphil** [PistonDevelopers/imageproc#97]
This might help: http://members.chello.at/~easyfilter/bresenham.html

**Comment by ArtemGr** [comment]
And then maybe answer this question:
http://stackoverflow.com/questions/34440429/draw-a-line-in-a-bitmap-possibly-with-piston  ; )

**Comment by dbrgn** [comment]
I added a simple Rust implementation of the Bresenham algorithm to the StackOverflow question: http://stackoverflow.com/a/34722500/284318

I'd be interested in giving this issue a stab. @theotherphil can you give me some starting point on where in the library I'd add something like that, and what stucts the function would operate on? Or in other words, what should the API look like?

**Comment by dbrgn** [comment]
Also, drawing a line with a thickness is probably just a specialization of filled rectangle drawing, right?

**Comment by dbrgn** [comment]
Sorry for the triple post, but I dug deeper in the docs and found some functions that are already present. Aren't these already covering the issue?

Drawing a line: [imageproc::drawing::draw_line_segment](http://docs.piston.rs/imageproc/imageproc/drawing/fn.draw_line_segment.html9
Drawing a thick line: [imageproc::drawing::draw_filled_rect](http://docs.piston.rs/imageproc/imageproc/drawing/fn.draw_filled_rect.html)

**Comment by dbrgn** [comment]
Can't get it to work though:

```rust
extern crate image;
extern crate imageproc;

use image::RgbImage;
use imageproc::rect::Rect;
use imageproc::drawing::draw_filled_rect;

fn main() {
    let mut img = RgbImage::new(256, 256);
    let thickness = 10;
    draw_filled_rect(&img, Rect::at(128, 10).of_size(118, thickness), [240, 240, 255]);
    img.save("output.png").unwrap();
}
```

Results in:

```
src/main.rs:16:5: 16:21 error: the trait `image::image::GenericImage` is not implemented for the type `image::buffer::ImageBuffer<image::color::Rgb<u8>, collections::vec::Vec<u8>>` [E0277]
src/main.rs:16     draw_filled_rect(&img, Rect::at(128, 10).of_size(118, thickness), [240, 240, 255]);
                   ^~~~~~~~~~~~~~~~
```

I'm not sure how I can use an `RgbImage` as a `GenericImage`.

**Comment by theotherphil** [comment]
There's only support for axis-aligned rectangles, so we still need to add code for slanted wide lines.

Your compiler error might just be because you're missing a "use image::GenericImage". If that doesn't work I'll check when I get home (about five hours from now).

More examples in the docs would be great, even if this approach doesn't end up solving the original problem.

**Comment by dbrgn** [comment]
> Your compiler error might just be because you're missing a "use image::GenericImage".

No, that didn't help.

**Comment by theotherphil** [comment]
Ah, the last parameter to draw_filled_rect should be Rgb([240, 240, 255]), instead of just the array.

**Comment by dbrgn** [comment]
That didn't solve the problem either :(

```rust
use image::{Rgb, RgbImage};
// ...

fn main() {
    // ...
    draw_filled_rect(&img, Rect::at(128, 10).of_size(118, thickness),
                     Rgb([240, 240, 255]));
    // ...
}
```

Same output as above.

**Comment by theotherphil** [comment]
Ok, I'll have a look when I get home.

**Comment by theotherphil** [comment]
Very odd. The following code works fine for me:

```rust
extern crate image;
extern crate imageproc;

use image::{
    Rgb,
    RgbImage
};

use imageproc::rect::Rect;
use imageproc::drawing::draw_filled_rect_mut;

fn main() {
    let mut img = RgbImage::new(256, 256);
    let thickness = 10;
    draw_filled_rect_mut(&mut img, Rect::at(128, 10).of_size(118, thickness), Rgb([240, 240, 255]));
    img.save("output.png").unwrap();
}
```

(Your code compiled too, but didn't do anything useful - I've replaced the use of draw_filled_rect with draw_filled_rect_mu so that we actually update the image.)

**Comment by theotherphil** [comment]
Are you trying this in a new project? If so, what does your Cargo.toml file look like?

**Comment by dbrgn** [comment]
I copy-pasted your code and still get the same errors :(

I'm using latest versions of both `image` and `imageproc`. Used `*` version specifiers for quick testing:

```
[package]
name = "img"
version = "0.1.0"
authors = ["..."]

[dependencies]
image = "*"
imageproc = "*"
```

**Comment by theotherphil** [comment]
For some reason image = "*" resulted in cargo producing a gigantic wall of complaints around not being able to find packages. However, with image = "0.3.15", imageproc = "*", I can compile and run successfully on latest stable, latest nightly, and nightly from the end of last year.

**Comment by dbrgn** [comment]
You're right, it works with image `0.3.15`. The current verision of `image` is `0.6.1` though. Could `imageproc` maybe be updated to support that version? And/or could/should important imports from `image` be re-exported?

**Comment by theotherphil** [comment]
Yes, it's pretty embarrassing that this doesn't work with more recent versions of image. I'd not noticed as I was using a fixed version of image. The breaking change looks to happen between 0.3.15 and 0.4.1.

**Comment by theotherphil** [comment]
This works with latest image:

```
extern crate image;
extern crate imageproc;

use image::{
    Rgb,
    RgbImage,
    GenericImage
};

fn constraints<I>(image: &I, color: I::Pixel)
    where I: GenericImage,
    I::Pixel: 'static
{
    let x = image.height();
    println!("{}", x);
}

fn main() {
    let img = RgbImage::new(256, 256);
    constraints(&img, Rgb([0, 0, 0]));
}
```

Where where I'm using the same bounds as the draw_filled_rect function. So I'm none the wiser as to what's going on.

**Comment by theotherphil** [comment]
@bvssvni Am I missing something obvious here?

**Comment by tafia** [comment]
When I do a `cargo clean` then `cargo build --verbose` it apparently compiles 2 versions of several crates
- png (0.4.1 and 0.3.1) 
- **image** (0.6.1 and 0.3.15)
- byteorder (0.4.2 and 0.3.13)
- bitflags (0.3.3 and 0.4.0)

If I force using image 0.6.1 (via `.cargo/config` file) the code works fine.
I suppose there is some conflict somewhere.

I also noted that there is a commit in image:
```
e7fd748 Fixed `NotEnoughData(..)` => `NotEnoughData`
```
which corresponds to the first warning messages I have.

**Comment by tafia** [comment]
@theotherphil I think the issue is in your Cargo.lock. 
It specifically mentions it uses image 0.3.15. You should update the dependencies it and push it to crates.io.


```
[[package]]
name = "imageproc"
version = "0.1.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "conv 0.3.1 (registry+https://github.com/rust-lang/crates.io-index)",
 "image 0.3.15 (registry+https://github.com/rust-lang/crates.io-index)",
 "itertools 0.4.6 (registry+https://github.com/rust-lang/crates.io-index)",
 "nalgebra 0.3.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "num 0.1.30 (registry+https://github.com/rust-lang/crates.io-index)",
 "quickcheck 0.2.24 (registry+https://github.com/rust-lang/crates.io-index)",
 "rand 0.3.13 (registry+https://github.com/rust-lang/crates.io-index)",
]
```

**Comment by korczis** [comment]
Any news? Any plans to fix/implement this?

I am looking for functionality similar to http://chunkypng.com/ 

Any plans to implement such functions/interface?

**Comment by theotherphil** [comment]
I had a quick look at the documentation for that library and couldn't see anything about drawing anti-aliased lines. Which functions specifically did you have in mind?

If you're just after PNG loading and saving then this is already supported by the [image](https://github.com/PistonDevelopers/image) library (whose image types and traits this library uses).

**Comment by korczis** [comment]
For example this method - http://www.rubydoc.info/github/wvanbergen/chunky_png/ChunkyPNG/Canvas/Drawing#line_xiaolin_wu-instance_method

**Comment by theotherphil** [comment]
Aha, thanks. I'll add that and the polygon function this weekend.

**Comment by korczis** [comment]
Wow. That will be really great. Thanks a lot.

**Comment by theotherphil** [comment]
I've made a start in https://github.com/PistonDevelopers/imageproc/pull/140. Take a look and let me know what you think. It's not as fast as it should be at the moment, but is hopefully correct. Is this good enough for your current use case? The initial aim of all the drawing functions was just to show points in interest on an image when debugging or verifying results of some vision algorithms, so none of them are particularly well written. This can certainly be improved though.

**Comment by korczis** [comment]
Works for me. Thanks.
